aws_lambda_powertools/tracing/tracer.py

```diff
@@ -280,7 +281,7 @@ def decorate(event, context):
 
         return decorate
 
-    def capture_method(self, method: Callable = None):
+    def capture_method(self, method: Callable = None):  # noqa: C901
```

Comment:
I'd like to avoid suppressing the "too complex" warning from flake8, but I'm not sure how to refactor this - open to suggestions!

Reply:
Now that we've grown to support generators, **what do you think of encapsulating the inner decorating logic into their own methods?**

That way we can have a cleaner `if/elif`, ease maintenance, and pass the linter -- There must be an even cleaner way if we think hard, but this is good enough (tm) for now IMHO.

```python
        if inspect.iscoroutinefunction(method):
            decorate = self._decorate_async_function(method=method)
        elif inspect.isgeneratorfunction(method):
            decorate = self._decorate_generator_function(method=method)
        elif hasattr(method, "__wrapped__") and inspect.isgeneratorfunction(method.__wrapped__):
            decorate = self._decorate_generator_function_with_context_manager(method=method)
        else:
            decorate = self._decorate_sync_function(method=method)
``` 

The private methods are basically a copy/paste of that logic we had:

```python
    def _decorate_async_function(self, method: Callable = None):
        method_name = f"{method.__name__}"

        @functools.wraps(method)
        async def decorate(*args, **kwargs):
            async with self.provider.in_subsegment_async(name=f"## {method_name}") as subsegment:
                try:
                    logger.debug(f"Calling method: {method_name}")
                    response = await method(*args, **kwargs)
                    self._add_response_as_metadata(function_name=method_name, data=response, subsegment=subsegment)
                except Exception as err:
                    logger.exception(f"Exception received from '{method_name}' method")
                    self._add_full_exception_as_metadata(function_name=method_name, error=err, subsegment=subsegment)
                    raise

                return response

        return decorate

    def _decorate_generator_function(self, method: Callable = None):
        method_name = f"{method.__name__}"

        @functools.wraps(method)
        def decorate(*args, **kwargs):
            with self.provider.in_subsegment(name=f"## {method_name}") as subsegment:
                try:
                    logger.debug(f"Calling method: {method_name}")
                    result = yield from method(*args, **kwargs)
                    self._add_response_as_metadata(function_name=method_name, data=result, subsegment=subsegment)
                except Exception as err:
                    logger.exception(f"Exception received from '{method_name}' method")
                    self._add_full_exception_as_metadata(function_name=method_name, error=err, subsegment=subsegment)
                    raise

                return result

        return decorate

    def _decorate_generator_function_with_context_manager(self, method: Callable = None):
        method_name = f"{method.__name__}"

        @functools.wraps(method)
        @contextlib.contextmanager
        def decorate(*args, **kwargs):
            with self.provider.in_subsegment(name=f"## {method_name}") as subsegment:
                try:
                    logger.debug(f"Calling method: {method_name}")
                    with method(*args, **kwargs) as return_val:
                        result = return_val
                    self._add_response_as_metadata(function_name=method_name, data=result, subsegment=subsegment)
                except Exception as err:
                    logger.exception(f"Exception received from '{method_name}' method")
                    self._add_full_exception_as_metadata(function_name=method_name, error=err, subsegment=subsegment)
                    raise

                yield result

        return decorate

    def _decorate_sync_function(self, method: Callable = None):
        method_name = f"{method.__name__}"

        @functools.wraps(method)
        def decorate(*args, **kwargs):
            with self.provider.in_subsegment(name=f"## {method_name}") as subsegment:
                try:
                    logger.debug(f"Calling method: {method_name}")
                    response = method(*args, **kwargs)
                    self._add_response_as_metadata(function_name=method_name, data=response, subsegment=subsegment)
                except Exception as err:
                    logger.exception(f"Exception received from '{method_name}' method")
                    self._add_full_exception_as_metadata(function_name=method_name, error=err, subsegment=subsegment)
                    raise

                return response

        return decorate

```

Reply:
Great idea, implemented as suggested